resourcex/traverse.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcex
+
+import (
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+// Traverse traverses a property value along the given property path,
+// invoking the given function for each property value it encounters.
+func Traverse(v resource.PropertyValue, p resource.PropertyPath, f func(resource.PropertyValue)) {
```

Comment:
I can't think of an example where we want to call some function for every property described by a path. Can you provide some motivating examples for this design (as opposed to calling the function at the value described by the path).

Reply:
The `Traverse` function is used by the `Extract` function to recover the unknownness and secretness of the extracted information. For example, an object value may itself be wrapped in a secret, contain fields that have values wrapped in secrets, etc.

Reply:
I might be confused. If that is the case and `.Metadata.Keys` is derived from `[1,2,3,4,5]`, doesn't that mean that `f` is called 5 times on the enclosing array?

---

`Extract` can call `traverse`, it doesn't necessitate making the function public.